### ISSUE TITLE
ActionCable: Repeated subscription attempts

### ISSUE THREAD

**Issue by sj26** [rails/rails#44652]
### Steps to reproduce

1. Create an actioncable subscription.
2. Await subscription confirmation.
3. Create another subscription with the same identifier.
4. Observe websocket showing subscription attempts every second.

I tried creating a small reproduction script, but involving views and action cable got too complicated. Instead, here's an app which replicates it:

https://github.com/sj26/action_cable_test

It creates two subscriptions with the same identifier, with a timeout to be sure that the first subscription has already been confirmed:

```erb
<div id="count"></div>
<%= javascript_tag do %>
  window.App.consumer.subscriptions.create({channel: "TestChannel"}, {
    received({ count }) {
      document.getElementById("count").innerText = count;
    }
  });
<% end %>

<div id="contents"></div>
<%= javascript_tag do %>
  setTimeout(function() {
    window.App.consumer.subscriptions.create({channel: "TestChannel"}, {
      received({ contents }) {
        document.getElementById("contents").innerText = contents;
      }
    });
  }, 1000)
<% end %>
```

Then a stream of subscription messages can be send in the websocket traffic:

<img width="1193" alt="image" src="https://user-images.githubusercontent.com/14028/157585817-14b2546d-0601-4faf-9334-e2615127be77.png">

because the guarantor considers the second subscription pending:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/14028/157586067-5ce7d08d-263a-4ea7-bfe4-9ab581f589b1.png">

because actioncable short circuits existing subscriptions, and so doesn't submit multiple confirmations:

https://github.com/rails/rails/blob/v7.0.2/actioncable/lib/action_cable/connection/subscriptions.rb#L32

### Use case

We have a react frontend, and we want to subscribe to channels in each of the leaf components which require dynamic updates directly. Sometimes this means multiple components will be interested in the same updates, and subscribe to the same identifiers. Trying to deduplicate these subscriptions somehow requires a lot more state management which seems more fragile and uneccessary. Previous versions of actioncable before the guarantor seem to work great for this use case, and the actioncable implemenation seem to consider subscription duplication. It's the addition of the subscription guarantor in #41581 to fix #38668 which seems to have created this issue. I think it would continue to work great with some gentle adjustment.

### Expected behavior

Subscribing to the same channel identifier multiple times should not constantly send subscription messages.

### Actual behavior

A continuous stream of subscription messages are sent from the browser and arrive and are squashed at the server without any confirmation message returning. This taxes both the browser client and the action cable server unnecessarily.

### System configuration
**Rails version**: 7.0.2

**Ruby version**: 3.1.0

### Potential solution

The simplest solution seems to be to follow the pattern established in the rest of Subscriptions and avoid re-subscribing subscriptions which already exist:

https://github.com/rails/rails/compare/main...sj26:actioncable-already-subscribed

It's a little confusing though. The "subscription" in the javascript environment does not necessarily have a 1:1 relationship with the "subscription" on the server. This is a good thing! But it'd be nice if they had different names.

**Comment by sj26** [comment]
Oh I guess the other solution would be to send a subscription confirmation from the actioncable server when a subscription is already in place, instead of swallowing it silently.

**Comment by salex** [comment]
I think I'm having the same or similar problem.

I tried to implement DHH's Rails 7 Blog/post demo in an application with some modifications. Mainly added a user and group that owned the post, converted the Slim, etc. I got everything working except the broadcast. I finally found the [Demo Code](https://github.com/paulfioravanti/rails7_steno_demo.git) (kinda hard to use the screencast to build code with all the jumping around) and found I had a DOM id problem. I worked in development. But then deployed it to a staging server and BOOM.

If I tried to go to Discussions(post) I got a Rails 502 error. Clicking refresh would bring up the post. Trying to open another browser window would get the rails error again on the first click, the open it. 

I would not stream and I also had some javascript (stimulus) that was supposed to highlight some markup but it just had a blank field. In other words, nothing worked.

I took out the streaming and redeployed and all other areas worked fine, just no streaming.

Here are about 20 lines each of rails, nginx and puma error logs,

[turbo.err.txt](https://github.com/rails/rails/files/8358266/turbo.err.txt)

Beyond my pay grade!

**Comment by sj26** [comment]
@salex that looks unrelated to the issue I'm describing here.

**Comment by sj26** [comment]
@matthewd I know you're pretty intimately familiar with action cable, do you have any advice here?

**Comment by sj26** [comment]
This is still reproducible.

**Comment by kevinhq** [comment]
any update on this? i have same problem to solve. multiple chat rooms, when I moved to another room, then go back to the first one or previous room, it creates another subscription with duplicated identifier:

![Screenshot from 2022-09-06 15-35-22](https://user-images.githubusercontent.com/1758073/188588917-0bd852d0-9d28-422d-8b51-da54817c1bab.png)

**Comment by kevinhq** [comment]
Update for whoever had same problem with same scenario with multiple chat rooms: I ended up using javascript to prevent creating/subscribing to server when the existing subscription already existed:

```
var arr_of_identifiers = consumer.subscriptions.subscriptions.map(s => {
      return s.identifier
    });
    var is_subscribed = false;
    for (const identifier of arr_of_identifiers) {
      if(identifier.includes(chatRoomId)) {
        is_subscribed = true;
        break;
      }
    }
    console.log(is_subscribed);
    if(is_subscribed == false) {
      // subscribe to server
    }
```

**Comment by lirimkrosa** [comment]
Check if cable is subscribed to specific channel.
```
  const isSubscribed = cableConsumer?.subscriptions?.subscriptions
    ?.map((x) => JSON.parse(x.identifier)?.channel)
    ?.includes('SUBSCRIBED_CHANNEL')
```

**Comment by fullc0ntr0l** [comment]
Have the same issue. Looking at the source, cable server expects only one subscription for the same identifier. It makes a little more sense if you add unsubscribe action into the context.
So you just created two subscriptions to the same channel, it uses the same identifier. What happens if you unsubscribe one of those subscription and remove it from client consumer? There is still one active subscription left on the client but the server will not send messages anymore.

**Comment by sj26** [comment]
Sorry I don't have any updates. I do think this is still an issue. We carry a patch for it in our codebase. But I don't have capacity to push it to completion myself. If someone else is keen, please let me know :pray:

**Comment by fullc0ntr0l** [comment]
Using this workaround for now

```ts
const subscription = consumer.subscriptions.create(...createArgs)
const subscriptionExists = consumer.subscriptions["findAll"](subscription.identifier).length > 1
if (subscriptionExists) {
    consumer.subscriptions["confirmSubscription"](subscription.identifier)
}
```

**Comment by sj26** [comment]
Yeah, that's roughly the patch we carry, as proposed in #44653 🙌

**Comment by x1wins** [comment]
<img width="601" alt="image" src="https://github.com/rails/rails/assets/982291/54ad1317-09b9-4427-8ecb-cfce66ae6b2e">

It's still happen.
Why ActionCable server don't send 'subscription confirm'?

Gemfile
```
ruby "3.2.2"
gem "rails", "~> 7.0.8"
```

Gemfile.lock
```
rails (7.0.8)
      actioncable (= 7.0.8)
```

**Comment by nicowenterodt** [comment]
I see something similar with actioncable (7.1.3.2). 

When I have a `turbo_stream_from` tag in place to connect my client to a stream and the client then navigates forth and back (using turbo-drive) to other pages where this tag is also in present, I see multiple `unsubscribes` and `subscribes` in my websocket monitor which occasionally leads to an unsubscribe from a stream which should be still connected.

When "hard-reloading" the page the websocket subscription to that stream is established as expected. So this only happens when a client navigates through the page.

This leads to the websocket not receiving messages anymore because the stream has been unsubscribed which leads to:

```
Turbo::StreamsChannel stopped streaming from ....
```

Occasionally throwing:
```
Could not execute command from ({"command"=>"unsubscribe" ....})
[RuntimeError - Unable to find subscription with identifier: ... ]
```

Imho there should be no unsubscribe at all when turbo-drive checks there is the same stream present again on the page the client navigates to.

I'm still trying to understand, however this seems related. Feels a bit like a race-condition.

**Comment by kulbirsaini** [comment]
This caused much pain for me today. Thank you for suggested hacks. I am using this for now.

```javascript

function createSubscription () {
    window.actionCableConsumer = window.actionCableConsumer || createConsumer()
    const subscription = window.actionCableConsumer.subscriptions.create(channelOptions)

    // Check if we are already subscribed to this channel. This is to avoid browser banging the server with subscribe requests
    // https://github.com/rails/rails/issues/44652#issuecomment-1712780124
    if (window.actionCableConsumer.subscriptions["findAll"](subscription.identifier).length > 1) {
        window.actionCableConsumer.subscriptions["confirmSubscription"](subscription.identifier)
    }
    return subscription
}
```

**Comment by kulbirsaini** [comment]
The hack code I mentioned above is not a fix. I still keep running into this issue.

![Screenshot-2024-06-18-123958-pplRBrwT](https://github.com/rails/rails/assets/127315/493c7b8c-31dc-4a39-a8b8-afd4f61e8dea)

**Comment by kulbirsaini** [comment]
I did what many others did, migrated to AnyCable. I have not seen this issue happening since.

**Comment by zzak** [comment]
I've recently observed something similar to this, by simply trying to subscribe to a channel that doesn't exist. :thinking:

**Comment by themaleem** [comment]
> Maybe related: [hotwired/turbo-rails#173](https://github.com/hotwired/turbo-rails/issues/173)
> 
> TL;TR:
> 
> I see something similar with actioncable (7.1.3.2).
> 
> When I have a `turbo_stream_from` tag in place to connect my client to a stream and the client then navigates forth and back (using turbo-drive) to other pages where this tag is also in present, I see multiple `unsubscribes` and `subscribes` in my websocket monitor which occasionally leads to an unsubscribe from a stream which should be still connected.
> 
> When "hard-reloading" the page the websocket subscription to that stream is established as expected. So this only happens when a client navigates through the page.
> 
> This leads to the websocket not receiving messages anymore because the stream has been unsubscribed which leads to:
> 
> ```
> Turbo::StreamsChannel stopped streaming from ....
> ```
> 
> Occasionally throwing:
> 
> ```
> Could not execute command from ({"command"=>"unsubscribe" ....})
> [RuntimeError - Unable to find subscription with identifier: ... ]
> ```
> 
> Imho there should be no unsubscribe at all when turbo-drive checks there is the same stream present again on the page the client navigates to.
> 
> I'm still trying to understand, however this seems related. Feels a bit like a race-condition.

can confirm it's still happening with `turbo_stream_from`

**Comment by Michoels** [comment]
This issue is still occurring in Turbo 8.0.6 in Sep 2024.
I switched to AnyCable as well, and the problem went away.

I ideally want to stay on the Rails Way™, so I'll keep any eye on this issue in the future.

**Comment by KerLq** [comment]
Same problem over here.

I also switched to AnyCable and encountered no problems

**Comment by jeremyhaile** [comment]
We just started using @rails/actioncable - and ran into this immediately. We have multiple React components, each of which can create a subscription and unsubscribes when they are unloaded.

However - when multiple components subscribe to the same channel (probably listening for different events), they constantly send subscription requests to the server. I was really hoping actioncable handled this case, as it seems like a very common use-case.

**Comment by goose3228** [comment]
**TL;DR;** - once turbo-stream tag is present on a page, make sure it does not get recreated on hotwire or react/vue/etc updates.

While there is certainly an error in ActionCable itself (DDOS vector at least) with this subscriptions, i was able to fix my case without migrating to anything.
The problem is that if you have a view (`frame.html.erb`) like that:
```
<div id="frame_123">
  <div>
      Frame
      <div>Content</div>
  </div>
  <%= turbo_stream_from "stream_123" %>
</div>
```

Which is replaced later with turbo or something
```
Turbo::StreamsChannel.broadcast_replace_to("stream_123", target: "frame_123", partial: "frame")
```

Here turbo-stream tag itself gets replaced on every broadcast, and you probably won't notice this in development/testing environments, it works fine. However, certain clients may cause subscription loops in production, and, if you are doing frequent broadcasts, can even make your server run out of RAM.

**Comment by JulesDebeaumont** [comment]
Happened to me on every latest Rails 7 I worked on. I haven't tried on Rails 8 yet.
I fixed it by updating Redis (was 7.0.5 to 7.4.1) following the official Redis documentation.
Seems like I was missing the lsb-release and maybe some depedencies, it's not just about installing Redis, at least on Linux.

```
sudo apt-get install lsb-release curl gpg
curl -fsSL https://packages.redis.io/gpg | sudo gpg --dearmor -o /usr/share/keyrings/redis-archive-keyring.gpg
sudo chmod 644 /usr/share/keyrings/redis-archive-keyring.gpg
echo "deb [signed-by=/usr/share/keyrings/redis-archive-keyring.gpg] https://packages.redis.io/deb $(lsb_release -cs) main" | sudo tee /etc/apt/sources.list.d/redis.list
sudo apt-get update
sudo apt-get install redis
```

https://redis.io/docs/latest/operate/oss_and_stack/install/install-redis/install-redis-on-linux/

**Comment by KieranP** [comment]
Have just run into the same issue, I've been struggling with this for a day now. Interestingly, it seems to work locally, but when deployed to production, it doesn't work.

Debug lines show the connection is establishing (and able to identify by current_user from the session env it receives), the channel is getting initialized (using current_user from connection and thread_id from subscription arguments), and it gets all the way to the `stream_for([current_user, thread])` line in the `subscribed` method of the channel.

But for some reason, the `stream_for` call is killing the subscription, and it gets into a loop, as seen below. As a result, the `connected` function of the actioncable.js subscription is never called. There are no errors on the rails side (websocket connection reaches rails connection/channel classes), and the browser only shows the following:

```
[ActionCable] Opening WebSocket, current state is null, subprotocols: actioncable-v1-json,actioncable-unsupported 1769983542074
[ActionCable] ConnectionMonitor started. stale threshold = 6 s 1769983542074
[ActionCable] WebSocket onopen event, using 'actioncable-v1-json' subprotocol 1769983542911
[ActionCable] ConnectionMonitor recorded connect 1769983542941
[ActionCable] SubscriptionGuarantor guaranteeing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983542941
[ActionCable] SubscriptionGuarantor resubscribing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983543442
[ActionCable] SubscriptionGuarantor already guaranteeing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983543442
[ActionCable] SubscriptionGuarantor resubscribing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983543942
[ActionCable] SubscriptionGuarantor already guaranteeing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983543943
[ActionCable] SubscriptionGuarantor resubscribing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983544445
[ActionCable] SubscriptionGuarantor already guaranteeing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983544445
[ActionCable] SubscriptionGuarantor resubscribing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983544946
[ActionCable] SubscriptionGuarantor already guaranteeing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983544946
[ActionCable] SubscriptionGuarantor resubscribing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983545448
[ActionCable] SubscriptionGuarantor already guaranteeing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983545448
[ActionCable] SubscriptionGuarantor resubscribing {"channel":"ChatChannel","thread_id":"697973f77693eef8ac8870fb"} 1769983545949
```

**Comment by ioquatix** [comment]
This should be fixed by https://github.com/rails/rails/pull/50979 -> duplicate subscriptions now raises an error.
